### ISSUE TITLE
adding charges to NEP

### ISSUE THREAD

**Issue by brucefan1983** [brucefan1983/GPUMD#299]
Many users hope to have this feature. 

How to set up the charges:
* Fixed charges?
  * let the user to specify?
  * let SNES to optimize?
 * Dynamic charges?
   * a separate NN is needed
   * need to study the literature

How to evaluate the interactions from the charges?
* the simple DSF method (damped shifted force); I like this
* the canonical Ewald summation; the golden standard

**Comment by ZKC19940412** [comment]
Hi, it is definitely a major step ahead to include charges. Here is a review from the HDNNP by Jörg Behler (https://pubs.acs.org/doi/abs/10.1021/acs.chemrev.0c00868) and page 15 - 20 goes over their latest scheme of incorporating charges  and might be useful for brainstorm the charge implementations for NEP. Just want to add a comment that while adding this new features, might be useful to do benchmarking on efficiency too? I love NEP because it is always the best in terms of balancing efficiency and accuracy.  I would love to see NEP is always the fastest code for large scale MD simulations.

**Comment by brucefan1983** [comment]
Thanks for the comments. This review is very helpful! I also hope to keep the superior speed of NEP, so I was thinking a non-Ewald way to evaluate the Coulomb forces.

**Comment by brucefan1983** [comment]
Aim to achieve this within GPUMD-v4.0.

**Comment by Ternity** [comment]
Recently, a new method about accelerating charge balance was developed by Jörg Behler et al., the titel of the article is [Accelerating fourth-generation machine learning potentials by quasi-linear scaling particle mesh charge equilibration](https://arxiv.org/abs/2403.02155). This may also be helpful for you.

**Comment by Ternity** [comment]
I also have a suggestion for using MLFF to predict charges:  Recently, Prof. Lei Li's research group at Southern University of Science and Technology has developed a method for self-consistent iterative prediction of atomic charge without reference to charge values. The paper web is: [https://pubs.acs.org/doi/10.1021/acs.jctc.3c01254?ref=pdf](https://pubs.acs.org/doi/10.1021/acs.jctc.3c01254?ref=pdf). 

With this approach, the charge of an atom can be predicted without considering which type of reference charge to use (e.g., bader, mulliken), thus avoiding the dilemma of choosing different charge reference types for different systems.

**Comment by brucefan1983** [comment]
Thank you very much for pointing out this very interesting paper!

**Comment by Ternity** [comment]
The code for paper "Charge-Optimized Electrostatic Interaction Atom-Centered Neural Network Algorithm" can be found in [PyAMFF](https://gitlab.com/pyamff/pyamff).

**Comment by brucefan1983** [comment]
Thank, I am actually experimenting with this method.

**Comment by brucefan1983** [comment]
Will continue to work on this after releasing GPUMD-4.0.

**Comment by Yuanwz00** [comment]
The current nep_charge utilizes an isotropic scalar charge, whereas adopting a tensor charge might offer a more accurate description of Coulomb interactions in anisotropic materials.

**Comment by erikfransson** [comment]
@brucefan1983 Is there any plans to also allow users to use NEPs with long-ranged forces from just the static nominal (user-specified) charges when training and running MD?

**Comment by brucefan1983** [comment]
@erikfransson Paul also mentioned this and I think it is not difficult to achieve this. I will do it when the requirements are clear.

**Comment by brucefan1983** [comment]
The major goals in this issue have been achieved upon the release of GPUMD-V4.6. https://github.com/brucefan1983/GPUMD/releases/tag/v4.6
